download/upgradeRecipe/upgradeRecipe6.2.adoc

```diff
@@ -0,0 +1,288 @@
+
+Before in `*.java`:
+[source, java]
+----
+public void addConstraintMatch(RuleContext kcontext, final int weight) {
+    ...
+    registerIntConstraintMatch(kcontext, 0, weight, new Runnable() {
+        public void run() {
+            ...
+        }
+    });
+}
+----
+
+After in `*.java`:
+[source, java]
+----
+public void addConstraintMatch(RuleContext kcontext, final int weight) {
+    ...
+    registerIntConstraintMatch(kcontext, 0, weight, new IntConstraintUndoListener() {
+        public void undo() {
+            ...
+        }
+    });
+}
+----
+
+====
+
+=== icon:exclamation-circle[] Major
+
+====
+If you have a custom `Move` implementation, now extract `AbstractMove`
+
+Before in `*.java`:
+[source, java]
+----
+public class CloudComputerChangeMove implements Move {...}
+----
+
+After in `*.java`:
+[source, java]
+----
+public class CloudComputerChangeMove extends AbstractMove {...}
+----
+
+====
+
+=== icon:info-circle[] Minor
+
+[.hidden-section]
+====
+The interface `Move` has a new method `getSimpleMoveTypeDescription()`.
+Extend `AbstractMove` so to avoid having to implement it.
+
+====
+
+=== icon:info-circle[] Minor
+
+[.hidden-section]
+====
+If you have a `@ValueRangeProvider` on an entity class (instead of the `Solution` class),
+then it's now safe to use the `<pillarChangeMoveSelector>` and `<pillarSwapMoveSelector>` as is.
+It's no longer needed to filter out swaps or changes
+which could put a value in an entity's variable that's not in its value range.
+
+Before in `*SolverConfig.xml` and `*BenchmarkConfig.xml`:
+[source, xml]
+----
+  <pillarChangeMoveSelector>
+    <filterClass>...ValidChangesOnlyPillarChangeMoveFilter</filterClass>
+  </pillarChangeMoveSelector>
+  <pillarSwapMoveSelector>
+    <filterClass>...ValidSwapsOnlyPillarSwapMoveFilter</filterClass>
+  </pillarSwapMoveSelector>
+----
+
+After in `*SolverConfig.xml` and `*BenchmarkConfig.xml`:
+[source, xml]
+----
+  <pillarChangeMoveSelector/>
+  <pillarSwapMoveSelector/>
+----
+
+====
+
+=== icon:info-circle[] Minor
+
+[.hidden-section]
+====
+`ConstraintMatchTotal` are now naturally compared by `scoreLevel`, then `constraintPackage`, then `constraintName`
+(instead of by `constraintPackage`, then `constraintName`, then `scoreLevel`) for readability.
+
+====
+
+=== icon:info-circle[] Minor
+
+[.hidden-section]
+====
+The optional method `IncrementalScoreCalculator.buildScoreCorruptionAnalysis(IncrementalScoreCalculator)` has been removed.
+Instead, to get a pretty score corruption analysis, implement the new interface `ConstraintMatchAwareIncrementalScoreCalculator`.
+This also enable your GUI to explain the score with an `IncrementalScoreCalculator`.
+See the example code in `MachineReassignmentIncrementalScoreCalculator.getConstraintMatchTotals()`.
+
+
+====
+
+== From 6.2.0.Beta1 to 6.2.0.Beta2
+
+
+=== icon:info-circle[] Minor
+
+[.hidden-section]
```

Comment:
This hidden-section is a lot of plumbing that we'll need to repeat n times (n being the number of upgrade recipe notes that we make), and it might be error-prone. I wonder if we can extract any of it in CSS and JavaScript?

IIRC, the section header "Minor" will create a div that spans that title and all paragraphs in it. But that won't work as the title would be hidden too... With CSS its possible to select the first child element of an element, but I doubt if it's possible to make that visible while the rest of the parent div isn't.